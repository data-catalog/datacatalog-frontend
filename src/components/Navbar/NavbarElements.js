import { FaBars } from 'react-icons/fa';
import { NavLink as Link } from 'react-router-dom';
import styled from 'styled-components';

export const RightWrapper = styled.div`
  display: flex;
  justify-content: space-between;
  width: 30%;
  align-items: center;
  margin-left: 24px;
`;

export const NavHome = styled(Link)`
  color: #fff;
  text-decoration: none;
  font-size: 200%;
  padding: 0 0 1% 0;
`;

export const Nav = styled.nav`
  background: #007382;
  height: 80px;
  display: flex;
  justify-content: space-between;
  padding: 0.5rem;
  z-index: 10;
  align-items: center;
`;

export const NavLink = styled(Link)`
  color: #fff;
  font-size: larger;
  display: flex;
  align-items: center;
  text-decoration: none;
  padding: 0 1rem;
  height: 100%;
  cursor: pointer;

  background-image: linear-gradient(120deg, #84fab0 0%, #8fd3f4 100%);
  background-repeat: no-repeat;
  background-size: 80% 0.2em;
  background-position: center bottom;
  transition: background-size 0.25s ease-in;
  &:hover {
    background-size: 100% 88%;
  }
`;

export const Bars = styled(FaBars)`
  display: none;
  color: #fff;
  @media screen and (max-width: 768px) {
    display: block;
    position: absolute;
    top: 0;
    right: 0;
    transform: translate(-100%, 75%);
    font-size: 1.8rem;
    cursor: pointer;
  }
`;

export const NavMenu = styled.div`
  display: flex;
  align-items: center;
  margin-right: 24px;
  @media screen and (max-width: 768px) {
    display: none;
  }
`;

export const NavBtn = styled.nav`
  display: flex;
  align-items: center;
  margin-right: 24px;
  @media screen and (max-width: 768px) {
    display: none;
  }
`;

export const NavBtnLink = styled(Link)`
  border-radius: 4px;
  background: #00B4CC;
  padding: 10px 22px;
  color: #fff;
  font-weight: bold;
  outline: none;
  border: none;
  cursor: pointer;
  transition: all 0.2s ease-in-out;
  text-decoration: none;
  margin-left: 24px;
  &:hover {
    transition: all 0.2s ease-in-out;
    background: #fff;
    color: #010606;
  }
`;