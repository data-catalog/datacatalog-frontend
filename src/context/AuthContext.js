import React, { createContext, useState } from 'react';
import Api from '../apis/api';

const AuthContext = createContext();

const AuthProvider = (props) => {
  const [user, setUser] = useState(JSON.parse(localStorage.getItem('user')));

  const login = async (data) => {
    const response = await Api.post('users/login', JSON.stringify(data));

    Api.defaults.headers.Authorization = `Bearer ${response.data.token}`;

    localStorage.setItem('access_token', response.data.token);
    localStorage.setItem('user', JSON.stringify(response.data.user));

    setUser(response.data.user);
  };

  const logout = () => {
    localStorage.removeItem('access_token');
    localStorage.removeItem('user');

    setUser(null);
  };

  return <AuthContext.Provider value={{ user, login, logout }} {...props} />;
};

function useAuth() {
  const context = React.useContext(AuthContext);
  if (context === undefined) {
    throw new Error('useAuth must be used within an AuthProvider');
  }
  return context;
}

export { AuthProvider, useAuth };
