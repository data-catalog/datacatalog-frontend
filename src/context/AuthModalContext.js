import React, { createContext, useState } from 'react';

const AuthModalContext = createContext();

const AuthModalProvider = (props) => {
  const [open, setOpen] = useState(false);
  const [variant, setVariant] = useState('login');

  return <AuthModalContext.Provider value={{ open, setOpen, variant, setVariant }} {...props} />;
};

export default AuthModalContext;
export { AuthModalProvider };
