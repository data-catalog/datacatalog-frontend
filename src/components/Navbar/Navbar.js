import React, { useState } from 'react';
import { BsFillPersonFill } from 'react-icons/bs';
import { useAuth } from '../../context/AuthContext';
import {
  Nav,
  NavLink,
  Bars,
  NavMenu,
  NavBtn,
  NavBtnLink,
  NavHome,
  RightWrapper,
  LeftWrapper,
  UsernameContainer,
} from './NavbarElements';
import Searchbar from '../SearchBar/SearchBar';
import SettingsMenu from '../SettingsMenu/SettingsMenu';

const SettingsButton = (props) => {
  const [open, setOpen] = useState(false);

  const { user, icon, children } = props;

  return (
    <>
      <LeftWrapper>
        <NavBtn>
          <NavBtnLink onClick={() => setOpen(!open)}> {icon} </NavBtnLink>
          {open && children(setOpen)}
        </NavBtn>
        {user && <UsernameContainer>{user.username}</UsernameContainer>}
      </LeftWrapper>
    </>
  );
};

const Navbar = () => {
  const { user } = useAuth();

  return (
    <>
      <Nav>
        <RightWrapper>
          <NavHome to="/">Data Catalog</NavHome>
          <Searchbar />
        </RightWrapper>
        <Bars />
        <NavMenu>
          <NavLink to="/about">Explore</NavLink>
          <NavLink to="/services">Library</NavLink>
        </NavMenu>
        <SettingsButton user={user} icon={<BsFillPersonFill />}>
          {(toggleMenu) => <SettingsMenu toggleMenu={toggleMenu} />}
        </SettingsButton>
      </Nav>
    </>
  );
};

export default Navbar;
export { SettingsButton };
