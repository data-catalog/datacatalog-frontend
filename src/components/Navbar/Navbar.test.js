import React from 'react';
import { mount, shallow } from 'enzyme';

import { BrowserRouter } from 'react-router-dom';
import Navbar, { SettingsButton } from './Navbar';
import { NavBtn, NavBtnLink } from './NavbarElements';
import { AuthProvider } from '../../context/AuthContext';

describe('NavMenu', () => {
  it('renders correctly', () => {
    const tree = shallow(
      <AuthProvider>
        <BrowserRouter>
          <Navbar />
        </BrowserRouter>
      </AuthProvider>
    );
    expect(tree).toMatchSnapshot();
  });
});

describe('SettingsButton', () => {
  it('toggles the settings menu when clicked', () => {
    const content = <p>Hidden text</p>;
    const wrapper = mount(<SettingsButton>{() => content}</SettingsButton>);

    // initially it should be closed
    expect(wrapper.find(NavBtn).contains(content)).toBe(false);

    // after click it should be open
    wrapper.find(NavBtnLink).at(0).simulate('click');
    expect(wrapper.find(NavBtn).contains(content)).toBe(true);

    // after another click it should be closed again
    wrapper.find(NavBtnLink).at(0).simulate('click');
    expect(wrapper.find(NavBtn).contains(content)).toBe(false);

    wrapper.unmount();
  });
});
