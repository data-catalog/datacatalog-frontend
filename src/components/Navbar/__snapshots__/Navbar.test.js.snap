// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavMenu renders correctly 1`] = `
<ContextProvider
  value={
    Object {
      "login": [Function],
      "logout": [Function],
      "user": null,
    }
  }
>
  <BrowserRouter>
    <Navbar />
  </BrowserRouter>
</ContextProvider>
`;
